internal/state/resolver/service_resolver_test.go

```diff
@@ -0,0 +1,180 @@
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/helpers"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/resolver"
+)
+
+func createEndpointSlice(name string, addresses []string, port int32, addressType discoveryV1.AddressType) *discoveryV1.EndpointSlice {
+	es := &discoveryV1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: "test",
+		},
+		AddressType: addressType,
+		Endpoints: []discoveryV1.Endpoint{
+			{
+				Addresses: addresses,
+				Conditions: discoveryV1.EndpointConditions{
+					Ready: helpers.GetBoolPointer(true),
+				},
+			},
+			{
+				Addresses: []string{"1.0.0.1", "1.0.0.2", "1.0.0.3"}, // these endpoints should be ignored because they are not ready
+				Conditions: discoveryV1.EndpointConditions{
+					Serving:     helpers.GetBoolPointer(true),
+					Terminating: helpers.GetBoolPointer(true),
+				},
+			},
+			{
+				Addresses:  []string{"2.0.0.1", "2.0.0.2", "2.0.0.3"},
+				Conditions: discoveryV1.EndpointConditions{
+					// nil conditions should be treated as not ready
+				},
+			},
+		},
+		Ports: []discoveryV1.EndpointPort{
+			{
+				Port: &port,
+			},
+		},
+	}
+
+	return es
+}
+
+func createFakeK8sClient(initObjs ...client.Object) (client.Client, error) {
+	scheme := runtime.NewScheme()
+	err := discoveryV1.AddToScheme(scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	fakeK8sClient := fake.NewClientBuilder().
+		WithScheme(scheme).
+		WithObjects(initObjs...).
+		Build()
+
+	return fakeK8sClient, nil
+}
+
+var _ = Describe("ServiceResolver", func() {
+	var (
+		addresses1        = []string{"9.0.0.1", "9.0.0.2"}
+		addresses2        = []string{"10.0.0.1", "10.0.0.2", "10.0.0.3"}
+		ipv6Addresses     = []string{"FE80:CD00:0:CDE:1257:0:211E:729C"}
+		diffPortAddresses = []string{"11.0.0.1", "11.0.0.2"}
+
+		svc = &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "test",
+				Name:      "svc",
+			},
+			Spec: v1.ServiceSpec{
+				Ports: []v1.ServicePort{
+					{
+						Port: 80,
+						TargetPort: intstr.IntOrString{
+							Type:   intstr.Int,
+							IntVal: 8080,
+						},
+					},
+					{
+						Port: 443,
+						TargetPort: intstr.IntOrString{
+							Type:   intstr.String,
+							StrVal: "8443",
+						},
+					},
+				},
+			},
+		}
+
+		slice1        = createEndpointSlice("svc-1", addresses1, 8080, discoveryV1.AddressTypeIPv4)
+		slice2        = createEndpointSlice("svc-2", addresses2, 8080, discoveryV1.AddressTypeIPv4)
+		sliceIPV6     = createEndpointSlice("svc-ipv6", ipv6Addresses, 8080, discoveryV1.AddressTypeIPv6)
+		slicePort8081 = createEndpointSlice("svc-diff-port", diffPortAddresses, 8081, discoveryV1.AddressTypeIPv4)
+	)
+
+	var (
+		fakeK8sClient   client.Client
+		serviceResolver resolver.ServiceResolver
+	)
+	Describe("Resolve", Ordered, func() {
+		BeforeAll(func() {
+			var err error
+			fakeK8sClient, err = createFakeK8sClient(slice1, slice2, sliceIPV6, slicePort8081)
```

Comment:
would it make sense to add a comment that "fakeK8sClient will return all EndpointSlices for the `List()` call done by the ServiceResolver"? 
Basically, the fact that the resolver uses `Service Name Index Field ` doesn't matter.